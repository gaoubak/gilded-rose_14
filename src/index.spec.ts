import runGoldenMaster from 'jest-golden-master';
import { GildedRose, Item } from './index';

describe('GildedRose constructor', () => {
  it('should instantiate with default items if none provided', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose();
      expect(gildedRose.items).toEqual([]);
    });
  });

  it('should instantiate with provided items', () => {
    runGoldenMaster(async () => {
      const items = [new Item('normal', 10, 20)];
      const gildedRose = new GildedRose(items);
      expect(gildedRose.items).toEqual(items);
    });
  });
});

describe('GildedRose', () => {

  it('should decrease quality and sellIn for a normal item', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('normal', 10, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(19);
      expect(items[0].sellIn).toBe(9);
    });
  });

  it('should increase quality for Aged Brie', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Aged Brie', 10, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(21);
    });
  });

  it('should not decrease quality below 0', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('normal', 10, 0)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(0);
    });
  });

  it('should handle Sulfuras correctly (no change)', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Sulfuras, Hand of Ragnaros', 10, 80)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(80);
      expect(items[0].sellIn).toBe(10);
    });
  });

  it('should increase quality for Backstage passes as concert approaches', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 12, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(21);
    });
  });

  it('should increase quality by 2 for Backstage passes with 10 days or less', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 10, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(22);
    });
  });

  it('should increase quality by 3 for Backstage passes with 5 days or less', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 5, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(23);
    });
  });

  it('should drop quality to 0 for Backstage passes after concert', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Backstage passes to a TAFKAL80ETC concert', 0, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(0);
    });
  });

  it('should not increase quality above 50 for Aged Brie', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Aged Brie', 10, 50)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(50);
    });
  });

  it('should degrade twice as fast for normal items once sell date has passed', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('normal', 0, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(18);
    });
  });

  it('should increase quality of Aged Brie by 1 when sellIn is negative', () => {
    runGoldenMaster(async () => {
      const gildedRose = new GildedRose([new Item('Aged Brie', -1, 20)]);
      const items = gildedRose.updateQuality();
      expect(items[0].quality).toBe(22);
    });
  });
});

test('should degrade Quality of "normal" item twice as fast after sell by date', () => {
  const item = new Item('normal', 0, 20); // Sell by date has passed
  const gildedRose = new GildedRose([item]);
  expect(item.quality).toBe(20);
  gildedRose.updateQuality();
  expect(item.quality).toBe(18); // Quality should decrease by 2
});


test('should never allow negative Quality', () => {
  const item = new Item('Normal Item', 5, 2); // Quality is already 0
  const gildedRose = new GildedRose([item]);
  gildedRose.updateQuality();
  expect(item.quality).toBeGreaterThanOrEqual(0); // Quality should remain 0
});

test('should increase Quality for "Aged Brie" as it gets older', () => {
  const item = new Item('Aged Brie', 5, 10);
  const gildedRose = new GildedRose([item]);
  gildedRose.updateQuality();
  expect(item.quality).toBeGreaterThanOrEqual(11); // Quality increases by 1
});
  
test('should never allow Quality to be more than 50', () => {
  const item = new Item('Aged Brie', 5, 40); // Quality is already 50
  const gildedRose = new GildedRose([item]);
  gildedRose.updateQuality();
  expect(item.quality).toBeLessThan(50); // Quality should remain 50
});
  
test('should never change SellIn or Quality for "Sulfuras"', () => {
  const item = new Item('Sulfuras, Hand of Ragnaros', 5, 80); // Legendary item
  const gildedRose = new GildedRose([item]);
  gildedRose.updateQuality();
  expect(item.quality).toBe(80); // Quality remains 80
  expect(item.sellIn).toBe(5); // SellIn remains 5
});


test('should increase Quality for "Backstage passes" according to SellIn value', () => {
  const item1 = new Item('Backstage passes to a TAFKAL80ETC concert', 10, 20);
  const gildedRose1 = new GildedRose([item1]);

  expect(item1.quality).toBe(20);
  
  gildedRose1.updateQuality();
  
  expect(item1.quality).toBe(22); // Quality should increase by 2
  
  const item2 = new Item('Backstage passes to a TAFKAL80ETC concert', 5, 20);
  const gildedRose2 = new GildedRose([item2]);
  
  expect(item2.quality).toBe(20);
  
  gildedRose2.updateQuality();
  
  expect(item2.quality).toBe(23); // Quality should increase by 3
  
  const item3 = new Item('Backstage passes to a TAFKAL80ETC concert', 0, 20);
  const gildedRose3 = new GildedRose([item3]);
  
  expect(item3.quality).toBe(20);
  
  gildedRose3.updateQuality();
  
  expect(item3.quality).toBe(0);
});

// thisdont work for now 
test('should increase Quality for "Backstage passes" as SellIn value approaches', () => {
  const item = new Item('Backstage passes to a TAFKAL80ETC concert', 9, 20);
  const gildedRose = new GildedRose([item]);
  expect(item.quality).toBe(20);
  gildedRose.updateQuality();
  expect(item.quality).toBe(21); // Quality should increase by 1 when SellIn is 10
});