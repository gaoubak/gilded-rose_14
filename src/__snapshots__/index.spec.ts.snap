// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GildedRose constructor should instantiate with default items if none provided 1`] = `[]`;

exports[`GildedRose constructor should instantiate with provided items 1`] = `[]`;

exports[`GildedRose should decrease quality and sellIn for a normal item 1`] = `
[
  "Before update:",
  "normal - SellIn: 10, Quality: 20",
  "Default item - Quality before: 20",
  "Default item - Quality after: 19",
  "Ensuring quality bounds for normal",
  "normal - Quality after ensuring bounds: 19",
  "After update:",
  "normal - SellIn: 9, Quality: 19",
]
`;

exports[`GildedRose should degrade twice as fast for normal items once sell date has passed 1`] = `
[
  "Before update:",
  "normal - SellIn: 0, Quality: 20",
  "Default item - Quality before: 20",
  "Default item - Quality after: 18",
  "Ensuring quality bounds for normal",
  "normal - Quality after ensuring bounds: 18",
  "After update:",
  "normal - SellIn: -1, Quality: 18",
]
`;

exports[`GildedRose should drop quality to 0 for Backstage passes after concert 1`] = `
[
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 0, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 0",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 0",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: -1, Quality: 0",
]
`;

exports[`GildedRose should handle Sulfuras correctly (no change) 1`] = `
[
  "Before update:",
  "Sulfuras, Hand of Ragnaros - SellIn: 10, Quality: 80",
  "After update:",
  "Sulfuras, Hand of Ragnaros - SellIn: 10, Quality: 80",
]
`;

exports[`GildedRose should increase quality by 2 for Backstage passes with 10 days or less 1`] = `
[
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 10, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 22",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 22",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 9, Quality: 22",
]
`;

exports[`GildedRose should increase quality by 3 for Backstage passes with 5 days or less 1`] = `
[
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 5, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 23",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 23",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 4, Quality: 23",
]
`;

exports[`GildedRose should increase quality for Aged Brie 1`] = `
[
  "Before update:",
  "Aged Brie - SellIn: 10, Quality: 20",
  "Aged Brie - Quality before: 20",
  "Aged Brie - Quality after: 21",
  "Ensuring quality bounds for Aged Brie",
  "Aged Brie - Quality after ensuring bounds: 21",
  "After update:",
  "Aged Brie - SellIn: 9, Quality: 21",
]
`;

exports[`GildedRose should increase quality for Backstage passes as concert approaches 1`] = `
[
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 12, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 21",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 21",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 11, Quality: 21",
]
`;

exports[`GildedRose should increase quality of Aged Brie by 1 when sellIn is negative 1`] = `
[
  "Before update:",
  "Aged Brie - SellIn: -1, Quality: 20",
  "Aged Brie - Quality before: 20",
  "Aged Brie - Quality after: 22",
  "Ensuring quality bounds for Aged Brie",
  "Aged Brie - Quality after ensuring bounds: 22",
  "After update:",
  "Aged Brie - SellIn: -2, Quality: 22",
]
`;

exports[`GildedRose should not decrease quality below 0 1`] = `
[
  "Before update:",
  "normal - SellIn: 10, Quality: 0",
  "Default item - Quality before: 0",
  "Default item - Quality after: -1",
  "Ensuring quality bounds for normal",
  "normal - Quality after ensuring bounds: 0",
  "After update:",
  "normal - SellIn: 9, Quality: 0",
]
`;

exports[`GildedRose should not increase quality above 50 for Aged Brie 1`] = `
[
  "Before update:",
  "Aged Brie - SellIn: 10, Quality: 50",
  "Aged Brie - Quality before: 50",
  "Aged Brie - Quality after: 51",
  "Ensuring quality bounds for Aged Brie",
  "Aged Brie - Quality after ensuring bounds: 50",
  "After update:",
  "Aged Brie - SellIn: 9, Quality: 50",
]
`;

exports[`should degrade Quality of "normal" item twice as fast after sell by date 1`] = `
[
  "Before update:",
  "normal - SellIn: 0, Quality: 20",
  "Default item - Quality before: 20",
  "Default item - Quality after: 18",
  "Ensuring quality bounds for normal",
  "normal - Quality after ensuring bounds: 18",
  "After update:",
  "normal - SellIn: -1, Quality: 18",
]
`;

exports[`should increase Quality for "Aged Brie" as it gets older 1`] = `
[
  "Before update:",
  "Aged Brie - SellIn: 5, Quality: 10",
  "Aged Brie - Quality before: 10",
  "Aged Brie - Quality after: 11",
  "Ensuring quality bounds for Aged Brie",
  "Aged Brie - Quality after ensuring bounds: 11",
  "After update:",
  "Aged Brie - SellIn: 4, Quality: 11",
]
`;

exports[`should increase Quality for "Backstage passes" according to SellIn value 1`] = `
[
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 10, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 22",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 22",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 9, Quality: 22",
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 5, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 23",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 23",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 4, Quality: 23",
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 0, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 0",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 0",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: -1, Quality: 0",
]
`;

exports[`should increase Quality for "Backstage passes" as SellIn value approaches 1`] = `
[
  "Before update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 10, Quality: 20",
  "Backstage Pass - Quality before: 20",
  "Backstage Pass - Quality after: 22",
  "Ensuring quality bounds for Backstage passes to a TAFKAL80ETC concert",
  "Backstage passes to a TAFKAL80ETC concert - Quality after ensuring bounds: 22",
  "After update:",
  "Backstage passes to a TAFKAL80ETC concert - SellIn: 9, Quality: 22",
]
`;

exports[`should never allow Quality to be more than 50 1`] = `
[
  "Before update:",
  "Aged Brie - SellIn: 5, Quality: 40",
  "Aged Brie - Quality before: 40",
  "Aged Brie - Quality after: 41",
  "Ensuring quality bounds for Aged Brie",
  "Aged Brie - Quality after ensuring bounds: 41",
  "After update:",
  "Aged Brie - SellIn: 4, Quality: 41",
]
`;

exports[`should never allow negative Quality 1`] = `
[
  "Before update:",
  "Normal Item - SellIn: 5, Quality: 2",
  "Default item - Quality before: 2",
  "Default item - Quality after: 1",
  "Ensuring quality bounds for Normal Item",
  "Normal Item - Quality after ensuring bounds: 1",
  "After update:",
  "Normal Item - SellIn: 4, Quality: 1",
]
`;

exports[`should never change SellIn or Quality for "Sulfuras" 1`] = `
[
  "Before update:",
  "Sulfuras, Hand of Ragnaros - SellIn: 5, Quality: 80",
  "After update:",
  "Sulfuras, Hand of Ragnaros - SellIn: 5, Quality: 80",
]
`;
